libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -428,34 +426,43 @@ VotesBundle VoteManager::getVotesBundleByRoundAndStep(uint64_t round, size_t ste
           }
           LOG(log_nf_) << "Found enough " << count << " votes at voted value " << voted_block_hash << " for round "
                        << round << " step " << step;
-          return VotesBundle(true, voted_block_hash, votes);
+          return VotesBundle(voted_block_hash, votes);
         }
       }
     }
   }
 
-  return VotesBundle();
+  return std::nullopt;
 }
 
 uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
+  std::vector<std::shared_ptr<Vote>> votes;
+
   SharedLock lock(verified_votes_access_);
-  for (auto rit = verified_votes_.rbegin(); rit != verified_votes_.rend(); ++rit) {
-    for (auto const& step : rit->second) {
-      if (step.first <= 3) {
-        continue;
+  for (auto round_rit = verified_votes_.rbegin(); round_rit != verified_votes_.rend(); ++round_rit) {
+    for (auto step_rit = round_rit->second.rbegin(); step_rit != round_rit->second.rend(); ++step_rit) {
+      if (step_rit->first <= 3) {
+        break;
       }
-      auto voted_block_hash_with_next_votes = getVotesBundleByRoundAndStep(rit->first, step.first, two_t_plus_one);
-      if (voted_block_hash_with_next_votes.enough) {
-        LOG(log_dg_) << "Found sufficient next votes in round " << rit->first << ", step " << step.first
-                     << ", PBFT 2t+1 " << two_t_plus_one;
-        // Update next votes
-        next_votes_manager_->updateNextVotes(voted_block_hash_with_next_votes.votes, two_t_plus_one);
-
-        return rit->first + 1;
+      for (auto const& voted_value : step_rit->second) {
+        if (voted_value.second.first >= two_t_plus_one) {
+          auto it = voted_value.second.second.begin();
+          size_t count = 0;
+
+          // Only copy 2t+1 votes
+          while (count < two_t_plus_one) {
+            votes.emplace_back(it->second);
+            count += it->second->getWeight().value();
+            it++;
+          }
+          LOG(log_nf_) << "Found enough " << count << " votes at voted value " << voted_value.first << " for round "
+                       << round_rit->first << " step " << step_rit->first;
+          next_votes_manager_->updateNextVotes(votes, two_t_plus_one);
+          return round_rit->first + 1;
+        }
```

Comment:
I don't understand the part. That's very similar with using getVotesBundleByRoundAndStep. What's benefits about the change?

Reply:
I think main idea is to not iterate from step 1,2 ... but reverse that, as the highest step has higher probability to have enough votes.